File: 501-find-mode-in-binary-search-tree/find-mode-in-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var findMode = function(root) {
 
    var ans = [];
    var count = 0;
    var max = 0;
    var prev = 10 ** 5 + 1;

    var rec = function(node) {
        if (node === null) {
            return;
        }

        rec(node.left);

        if (node.val === prev) {
            count++;
        } else {
            count = 1;
        }

        if (count === max) {
            ans.push(node.val);
        } else if (count > max) {
            ans = [node.val];
            max = count;
        }

        prev = node.val;

        rec(node.right);
    };

    rec(root);
    return ans;
};

File: 501-find-mode-in-binary-search-tree/find-mode-in-binary-search-tree.py
from collections import defaultdict as dict

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findMode(self, root: Optional[TreeNode]) -> List[int]:
        ans = []
        count = 0
        max_ = 0
        prev = 10**5 + 1

        def rec(root):
            if root is None:
                return
            nonlocal count, max_, prev, ans 
            rec(root.left)
            if prev == root.val:
                count += 1
            else:
                count = 1
            if count == max_:
                ans.append(root.val)
            elif count > max_:
                ans = [root.val]
                max_ = count
            prev = root.val
            rec(root.right)

        rec(root)
        return ans
